src/util/openfaas/index.js

```diff
@@ -230,58 +233,74 @@ async function setupFaasFunction(
 }
 
 const executeFaasFunction = async (
-  functionName,
+  name,
   events,
   versionId,
   libraryVersionIDs,
   testMode,
   trMetadata = {},
 ) => {
-  try {
-    logger.debug('[Faas] Invoking faas function');
+  logger.debug(`Executing faas function: ${name}`);
 
-    if (testMode) await awaitFunctionReadiness(functionName);
+  const startTime = new Date();
+  let errorRaised;
+
+  try {
+    if (testMode) await awaitFunctionReadiness(name);
+    return await invokeFunction(name, events);
 
-    const res = await invokeFunction(functionName, events);
-    logger.debug('[Faas] Invoked faas function');
-    return res;
   } catch (error) {
-    logger.error(`[Faas] Error while invoking ${functionName}: ${error.message}`);
+    logger.error(`Error while invoking ${name}: ${error.message}`);
+    errorRaised = error;
+
     if (
       error.statusCode === 404 &&
-      error.message.includes(`error finding function ${functionName}`)
+      error.message.includes(`error finding function ${name}`)
     ) {
-      removeFunctionFromCache(functionName);
+      removeFunctionFromCache(name);
       await setupFaasFunction(
-        functionName,
+        name,
         null,
         versionId,
         libraryVersionIDs,
         testMode,
         trMetadata,
       );
-      throw new RetryRequestError(`${functionName} not found`);
+      throw new RetryRequestError(`${name} not found`);
     }
 
     if (error.statusCode === 429) {
-      throw new RetryRequestError(`Rate limit exceeded for ${functionName}`);
+      throw new RetryRequestError(`Rate limit exceeded for ${name}`);
     }
 
     if (error.statusCode === 500 || error.statusCode === 503) {
       throw new RetryRequestError(error.message);
     }
 
     if (error.statusCode === 504) {
-      throw new RespStatusError('Timed out');
+      throw new RespStatusError(`${name} timed out`);
     }
 
     throw error;
   } finally {
+    // delete the function created, if it's called as part of testMode
     if (testMode) {
-      deleteFunction(functionName).catch((err) => {
-        logger.error(`[Faas] Error while deleting ${functionName}: ${err.message}`);
-      });
+      deleteFunction(name).catch((err) => 
+        logger.error(`[Faas] Error while deleting ${name}: ${err.message}`))
     }
+
+    // setup the tags for observability and then fire the stats
+    const tags = {
+      identifier: "openfaas",
+      testMode: testMode,
+      errored: errorRaised ? true : false,
+      statusCode: errorRaised ? errorRaised.statusCode : HTTP_STATUS_CODES.OK, // default statuscode is 200OK
+      ...events.length && events[0].metadata ? getMetadata(events[0].metadata) : {},
+      ...events.length && events[0].metadata ? getTransformationMetadata(events[0].metadata) : {},
```

Comment:
metadata is already passed as `trMetadata`. why not use it

Reply:
The reason not use it would be to prevent accidently updates to trMetadata in terms of adding more data and the stat stop firing. Better to construct all required labels within the function, so we exactly know what's happening.